File: src/main/java/main/puzzle/assembly/dxz/zdd/ZDDNodeTable.java
package main.puzzle.assembly.dxz.zdd;

import main.util.ThreadPoolManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 *
 * @author Bunnyspa
 */
public class ZDDNodeTable {

    private final Map<Integer, Set<ZDDNode>> map = new HashMap<>();

    @Nullable
    ZDDNode get(int i, ZDDNode l, ZDDNode h) {
        if (!map.containsKey(i)) {
            return null;
        }
        for (ZDDNode node : map.get(i)) {
            if (node.equals(i, l, h)) {
                return node;
            }
        }
        return null;
    }

    void add(@NotNull ZDDNode node) {
        ThreadPoolManager.getThreadPool().execute(() -> {
            int key = node.label;
            if (!map.containsKey(key)) {
                map.put(key, new HashSet<>());
            }
            map.get(key).add(node);
        });
    }

}

File: src/main/java/main/puzzle/assembly/dxz/zdd/ZDDNodeTable.kt
package main.puzzle.assembly.dxz.zdd

import main.util.ThreadPoolManager
import java.util.*

/**
 *
 * @author Bunnyspa
 */
class ZDDNodeTable {
    private val map: MutableMap<Int, MutableSet<ZDDNode>> = HashMap()
    operator fun get(i: Int, l: ZDDNode?, h: ZDDNode): ZDDNode? {
        if (!map.containsKey(i)) {
            return null
        }
        for (node in map[i]!!) {
            if (node.equals(i, l, h)) {
                return node
            }
        }
        return null
    }

    fun add(node: ZDDNode) {
        ThreadPoolManager.threadPool.execute {
            val key = node.label
            if (!map.containsKey(key)) {
                map[key] = HashSet()
            }
            map[key]!!.add(node)
        }
    }
}